crates/mun_compiler/src/diagnostics_snippets.rs

```diff
@@ -0,0 +1,372 @@
+    let first_definition_location = syntax_node_ptr_location(diagnostic.first_definition, &parse);
+    let definition_location = syntax_node_ptr_location(diagnostic.definition, &parse);
+
+    let duplication_object_type =
+        if matches!(diagnostic.first_definition.kind(), SyntaxKind::STRUCT_DEF)
+            && matches!(diagnostic.definition.kind(), SyntaxKind::STRUCT_DEF)
+        {
+            "type"
+        } else {
+            "value"
+        };
+
+    SnippetBuilder::new()
+        .title(
+            AnnotationBuilder::new(AnnotationType::Error)
+                .label(diagnostic.message())
+                .build(),
+        )
+        .slice(
+            SliceBuilder::new(true)
+                .origin(relative_file_path.to_string())
+                // First definition
+                .source_annotation(
+                    text_range_to_tuple(first_definition_location),
+                    format!(
+                        "previous definition of the {} `{}` here",
+                        duplication_object_type, diagnostic.name
+                    ),
+                    AnnotationType::Warning,
+                )
+                // Second definition
+                .source_annotation(
+                    text_range_to_tuple(definition_location),
+                    format!("`{}` redefined here", diagnostic.name),
+                    AnnotationType::Error,
+                )
+                .build(&source_code, source_code_len, &line_index),
+        )
+        .footer(
+            AnnotationBuilder::new(AnnotationType::Note)
+                .label(format!(
+                    "`{}` must be defined only once in the {} namespace of this module",
+                    diagnostic.name, duplication_object_type
+                ))
+                .build(),
+        )
+        .build()
+}
+
+pub(crate) fn possibly_uninitialized_variable_error(
+    diagnostic: &mun_hir::diagnostics::PossiblyUninitializedVariable,
+    _: &impl HirDatabase,
+    parse: &Parse<SourceFile>,
+    relative_file_path: &str,
+    source_code: &str,
+    source_code_len: usize,
+    line_index: &Arc<LineIndex>,
+) -> Snippet {
+    let variable_name = diagnostic.pat.to_node(&parse.syntax_node()).text();
+
+    SnippetBuilder::new()
+        .title(
+            AnnotationBuilder::new(AnnotationType::Error)
+                .label(format!("{}: `{}`", diagnostic.message(), variable_name))
+                .build(),
+        )
+        .slice(
+            SliceBuilder::new(true)
+                .origin(relative_file_path.to_string())
+                .source_annotation(
+                    text_range_to_tuple(diagnostic.highlight_range()),
+                    format!("use of possibly-uninitialized `{}`", variable_name),
+                    AnnotationType::Error,
+                )
+                .build(&source_code, source_code_len, &line_index),
+        )
+        .build()
+}
+
+pub(crate) fn access_unknown_field_error(
```

Comment:
If there is this much code duplication, my intuition is that there should be a smarter/more concise way of doing this. I'd have to investigate to see what that method is.